### PR TITLE
Fix readiness probe when sidecar injector is disabled

### LINKED ISSUE #45990: Istiod never becomes ready when sidecar injection is disabled
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Upgraded Istio from 1.16.5 to 1.17.3 and now istiod no longer becomes ready.  Upon further inspection istiod is outputting:
`warn	sidecar injector is not ready`, which when looking into the code is due to the combined readiness endpoint.  Our Istio is configured with injection disabled, so this should not be part of the readiness check.

It appears the changes in: https://github.com/istio/istio/commit/b0ed8cb185c27fcd9bc008c0f04f84f6120d1805, did not either:
1. Conditionally add the probe
2. Set the probe to ready if injection was not enabled when `Server.initSidecarInjector()` exits early due to injection not being enabled/configured.

I would propose adding the readiness to true within `Server.initSidecarInjector()` on early exit as that should catch the various (non-error) reasons why injection may not be enabled without duplicating logic.

### Version

```Text
istioctl version     
client version: 1.18.0
control plane version: 1.16.5 (note this is last ready version not the version with the issue)
data plane version: 1.17.3 (6 proxies)
```

### Additional Information

_No response_

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### PR DESCRIPTION
**Pull request by arsenetar**
Make sure the readiness probe returns ready when the sidecar injector is not used, or not configured.

Fix #45990

**Please provide a description of this PR:**
Resolves the bug reported in #45990 by making sure the readiness probe for the sidecar injector is true when the injector is not used.  This prevents istiod from getting stuck in the non-ready state.

### CODE REVIEW

**Comment by arsenetar** [comment]
Apologies if I have missed some required PR step here or marked something incorrectly. 

I wasn't exactly sure if this need a releasenotes file, the docs seem to suggest it should, but other PRs like this don't seem to have it.  Additionally, I was not sure what `area` to use for the note if it does require one.

**Review by hanxiaop — APPROVED** [review]

**Comment by ericvn** [comment]
@arsenetar Can you satisfy the CLA. Also, do we want to cherry-pick this to 1.18 since the issue shows the problem there?

**Comment by arsenetar** [comment]
@ericvn CLA is in process, might have run into an issue with not getting the email for the corporate signing.

**Review by hzxuzhonghu — APPROVED** [review]

**Review by hzxuzhonghu — COMMENTED** [review]
one nit: you can set ready after `initSidecarInjector` retiurns nil, then you donot need to set everywhere. I will put a hold now

**Comment by arsenetar** [comment]
@hzxuzhonghu Refactored.

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -319,6 +319,7 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
 		if err != nil {
 			return nil, fmt.Errorf("error initializing sidecar injector: %v", err)
 		}
+		s.readinessFlags.sidecarInjectorReady.Store(true)
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:322]: YOu should add 
if wh == nil{
    s.readinessFlags.sidecarInjectorReady.Store(true)
}

Comment (arsenetar) [reply pilot/pkg/bootstrap/server.go:322]: Fairly certain wrapping the readiness in the suggested conditional will prevent the cases where the injector is used from ever becoming ready as this is the only location where the readiness is marked as true.

**Review by hzxuzhonghu — APPROVED** [review]

**Comment by hanxiaop** [comment]
@hzxuzhonghu Can the DNM label be removed? @arsenetar Can you sign the CLA? Thanks

**Comment by arsenetar** [comment]
We had a hold up with the corporate CLA, got permission to do a individual contributor on this contribution instead.  Sorry for the delay.

**Comment by hanxiaop** [comment]
It looks like the cherry-pick bot didn't work?

**Comment by zetaj** [comment]
Hi, we are running into the same issue as https://github.com/istio/istio/issues/45990 mentioned.
(Context: we are upgrading Istio to 1.18.2)
I have checked the branch 1.18.2 and do not see the fix included.
Wondering what is the ETA for releasing it. Thanks.

**Comment by hanxiaop** [comment]
> Hi, we are running into the same issue as #45990 mentioned. (Context: we are upgrading Istio to 1.18.2) I have checked the branch 1.18.2 and do not see the fix included. Wondering what is the ETA for releasing it. Thanks.

@zetaj 1.18.2 includes only CVE fixes and 1.18.3 will include the readiness fix. Should be available in ~3 weeks.

**Pull request merged**